pkg/cmd/dcos.go

```diff
@@ -179,6 +179,21 @@ func invokePlugin(ctx api.Context, cmd plugin.Command, args []string) error {
 		} else if cluster.TLS().RootCAsPath != "" {
 			execCmd.Env = append(execCmd.Env, "DCOS_TLS_CA_PATH="+cluster.TLS().RootCAsPath)
 		}
+
+		// Create env entries based on the subcommand config. For example if `hello.world = "foo"`
+		// is present in the config file and the `hello` command is invoked, an environment variable
+		// `DCOS_HELLO_WORLD=foo` will be passed when invoking `dcos hello`.
+		if cmdConfig, ok := cluster.Config().ToMap()[cmd.Name].(map[string]interface{}); ok {
+			for key, val := range cmdConfig {
+				entry := fmt.Sprintf(
+					"DCOS_%s_%s=%v",
+					strings.ToUpper(cmd.Name),
+					strings.ToUpper(key),
+					val,
+				)
+				execCmd.Env = append(execCmd.Env, entry)
+			}
+		}
```

Comment:
I'd like to extract above code to a function that calculates env and add unit tests for it

```diff
diff --git a/pkg/cmd/dcos.go b/pkg/cmd/dcos.go
index ac4c967..22d2a87 100644
--- a/pkg/cmd/dcos.go
+++ b/pkg/cmd/dcos.go
@@ -159,25 +159,43 @@ func invokePlugin(ctx api.Context, cmd plugin.Command, args []string) error {
 	execCmd.Stderr = ctx.ErrOut()
 	execCmd.Stdin = ctx.Input()
 
-	execCmd.Env = append(os.Environ(), "DCOS_CLI_EXECUTABLE_PATH="+executablePath)
+	execCmd.Env = append(execCmd.Env, calculateEnv(ctx, executablePath, cmd)...)
 
+	err = execCmd.Run()
+	if err != nil {
+		// Because we're silencing errors through Cobra, we need to print this separately.
+		ctx.Logger().Debug(err)
+		// When the plugin command exits with a non-zero code, the main CLI process
+		// exit code should be the same. ExitCode() returns -1 if the process
+		// hasn't exited or was terminated by a signal thus we have to check.
+		//
+		// See https://jira.mesosphere.com/browse/DCOS_OSS-4399
+		if exitCode := execCmd.ProcessState.ExitCode(); exitCode > 0 {
+			os.Exit(exitCode)
+		}
+
+	}
+	return err
+}
+
+func calculateEnv(ctx api.Context, executablePath string, cmd plugin.Command) (env []string) {
+	env = append(os.Environ(), "DCOS_CLI_EXECUTABLE_PATH="+executablePath)
 	switch ctx.Logger().Level {
 	case logrus.DebugLevel:
-		execCmd.Env = append(execCmd.Env, "DCOS_VERBOSITY=2", "DCOS_LOG_LEVEL=debug")
+		env = append(env, "DCOS_VERBOSITY=2", "DCOS_LOG_LEVEL=debug")
 	case logrus.InfoLevel:
-		execCmd.Env = append(execCmd.Env, "DCOS_VERBOSITY=1", "DCOS_LOG_LEVEL=info")
+		env = append(env, "DCOS_VERBOSITY=1", "DCOS_LOG_LEVEL=info")
 	}
-
 	// Pass cluster specific env variables when a cluster is attached.
 	if cluster, err := ctx.Cluster(); err == nil {
-		execCmd.Env = append(execCmd.Env, "DCOS_URL="+cluster.URL())
-		execCmd.Env = append(execCmd.Env, "DCOS_ACS_TOKEN="+cluster.ACSToken())
+		env = append(env, "DCOS_URL="+cluster.URL())
+		env = append(env, "DCOS_ACS_TOKEN="+cluster.ACSToken())
 
 		insecure := cluster.TLS().Insecure || strings.HasPrefix(cluster.URL(), "http://")
 		if insecure {
-			execCmd.Env = append(execCmd.Env, "DCOS_TLS_INSECURE=1")
+			env = append(env, "DCOS_TLS_INSECURE=1")
 		} else if cluster.TLS().RootCAsPath != "" {
-			execCmd.Env = append(execCmd.Env, "DCOS_TLS_CA_PATH="+cluster.TLS().RootCAsPath)
+			env = append(env, "DCOS_TLS_CA_PATH="+cluster.TLS().RootCAsPath)
 		}
 
 		// Create env entries based on the subcommand config. For example if `hello.world = "foo"`
@@ -191,26 +209,11 @@ func invokePlugin(ctx api.Context, cmd plugin.Command, args []string) error {
 					strings.ToUpper(key),
 					val,
 				)
-				execCmd.Env = append(execCmd.Env, entry)
+				env = append(env, entry)
 			}
 		}
 	}
-
-	err = execCmd.Run()
-	if err != nil {
-		// Because we're silencing errors through Cobra, we need to print this separately.
-		ctx.Logger().Debug(err)
-		// When the plugin command exits with a non-zero code, the main CLI process
-		// exit code should be the same. ExitCode() returns -1 if the process
-		// hasn't exited or was terminated by a signal thus we have to check.
-		//
-		// See https://jira.mesosphere.com/browse/DCOS_OSS-4399
-		if exitCode := execCmd.ProcessState.ExitCode(); exitCode > 0 {
-			os.Exit(exitCode)
-		}
-
-	}
-	return err
+	return env
 }
 
 // help menu template that follow UX styleguide.
```

Reply:
I've added unit tests in the latest commits.